Resolve DNS Hosted Zone ID while building IAM policy

zytek opened the pull request:
Fixes #1949 by resolving DNS Zone prior to running IAM builder.

Also small refactor to remove code duplication (thanks @yissachar )

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1969)
<!-- Reviewable:end -->

zytek commented:
rebased to pull in fixes for CI

justinsb commented on pkg/model/iam/iam_builder.go:
```diff
@@ -212,6 +175,26 @@ func (b *IAMPolicyBuilder) BuildAWSIAMPolicy() (*IAMPolicy, error) {
 		}
 	}
 
+	p.Statement = append(p.Statement, &IAMStatement{
```
Moving this is great - the code repetition is less acceptable given the route53 policies are no longer trivial.  But this is what is breaking the cloudformation test, because the order of the rules has changed.

A few options I think:

* Don't do the refactor
* Fix the test (it shouldn't be too bad, just rearranging the sections mostly)
* Pull this into a function and call it twice

justinsb commented on upup/pkg/fi/cloudup/apply_cluster.go:
```diff
@@ -346,6 +346,8 @@ func (c *ApplyClusterCmd) Run() error {
 	if err != nil {
 		return err
 	}
+	dnszone, _ := findZone(cluster, cloud) // any errors would be already caught in validateDNS
```
Can we check the error - even if it shouldn't happen?  We do make calls to the cloud APIs, so I think it is possible that things fail the second time (e.g. rate limits).

justinsb commented:
LGTM.  We have to get the tests passing, and also I always worry about not checking errors, but generally great - thank you - nice fix!

justinsb commented:
(sorry, hit the wrong button!)

zytek commented:
* fixed cloudformation template
* fixed error checking when calling findZone (yeah, that wasn't smart to ignore error in function calling external APIs)

CI still fails on my end due to
```
Boilerplate header is wrong for: /Users/zytek/go/src/k8s.io/kops/out/production/kops/upup/models/bindata.go
Execute hack/update-header.sh to update headers
```

but this seems unrelated to my changes. Fails even after rebasing to current master. 


thx @justinsb !

justinsb commented:
@k8s-bot ok to test
